keps/sig-api-machinery/1040-priority-and-fairness/README.md

```diff
@@ -638,24 +640,107 @@ always dispatched immediately.  Following is how the other requests
+the API and may be any value between zero and
+`assuredConcurrencyShares` inclusive (default is zero).  Each
+apiserver allocates NominalConcurrencyLimits in proportion to
+`assuredConcurrencyShares` and LendableConcurrncyLimit in
+corresponding propotion:
+
+```
+NominalConcurrencyLimit(i)  = ceil( SCL * assuredConcurrencyShares(i)  / sum_assured )
+LendableConcurrencyLimit(i) = ceil( SCL * lendableConcurrencyShares(i) / sum_assured )
+NonLendableConcurrencyLimit(i) = NominalConcurrencyLimit(i) - LendableConcurrencyLimit(i)
+sum_assured  = sum[priority levels k] assuredConcurrencyShares(k)
+```
+
+where SCL is the apiserver's concurrency limit.
+
+Borrowing is further limited by a practical consideration: we do not
+want a global mutex covering all dispatching.  Aside from borrowing,
+dispatching from one priority level is done independently from
+dispatching at another.  Borrowing is allowed in just one direction
+(higher may borrow from lower, and lower does not actively lend to
+higher) and in a very limited quantity: an attempt to dispatch for one
+priority level will consider borrowing from just one other priority
+level (if there is any lower priority level at all).
+
+A request can be dispatched exactly at a non-exempt priority level
+when either there are no requests executing at that priority level or
+the number of seats needed by that request is no greater than the
+number of unused seats at that priority level.  The number of unused
+seats at a given priority level is that level's
+NominalConcurrencyLimit minus the number of seats used by requests
+executing at that priority level (dispatched from that priority level
+and higher ones).
+
+There are two sorts of times when dispatching to a non-empty priority
+level is considered: when a request arrives, and when a request
+releases the seats it was occupying (which is not the same as when the
+request finishes from the client's point of view, see below about
+WATCH requests).
+
+At each of these sorts of moments, as many requests are dispatched
+exactly at the same priority level as possible.  The next request to
+consider dispatching is chosen by using the Fair Queuing for Server
+Requests algorithm below to choose a queue at that priority level, and
+the request at the head of that queue is considered.  If (a) no
+requests can be dispatched exactly at that priority level at that
+moment, (b) there are non-empty queues at that level, and (c) there
+are lower non-exempt priority levels, then the request at the head of
+the chosen queue is considered for dispatch at one of the lower
+priority levels.  The particular lower priority level considered is
+drawn at random from the lower ones, in proportion to their
+LendableConcurrencyLimit (we use a static value so that the drawing
+can be done without acquiring mutexes).  The request is executed at
+the chosen lower level (occupying some of its seats) if the request
+can be dispatched exactly at that level according to the rule above.
+
```

Comment:
One thing that I'm missing is the fact that what changes now is:

before:
(1) when a request releases its seats after processing, we look for requests to dispatch only at its PL

after:
(2) when a request releases its seats after processing, it may not only mean that some other request at the same PL is now dispatchable, but it may also mean that requests from some higher PL become dispatchable now

So I think that we need to decide:
(a) what do we prioritize [I would say that if there is a request from a given PL that we can now dispatch, we dispatch this one]
(b) if we can't dispatch request from the same PL, do we try to dispatch requests from higher PLs? And if so, in what order? [I would say that we should do that, if and only if our PL is empty (no waiting requests) and do that starting from the highest priority PL]

Reply:
Revised to explicitly address this.

Reply:
Remember that the queue being dispatched _from_ is in a priority level, so I think the lock on that priority level has to be held until a dispatching decision is made.

Reply:
So what I'm suggesting is:
- let's say we released all seats on a given PL and we have an empty queue there
- we forget about the queue
- but we start from the highest priority level and try to dispatch from those queues (just optimistically assuming that our queue will remain to have free seats) - yes you will have to take a lock on our initial queue again, but that sounds acceptable if there are requests to dispatch in higher PLs

Reply:
> yes you will have to take a lock on our initial queue again, but that sounds acceptable if there are requests to dispatch in higher PLs

This matches what I would expect.